### ISSUE TITLE
Cancellation behavior on Windows and how to improve it?

### ISSUE THREAD

**Issue by amitsaha** [buildkite/agent#794]
On Windows, job cancellation is:

```
# process\process.go
if runtime.GOOS == "windows" {
    // Sending Interrupt on Windows is not implemented.
    // https://golang.org/src/os/exec.go?s=3842:3884#L110
   err = exec.Command("CMD", "/C", "TASKKILL", "/F", "/PID", strconv.Itoa(p.Pid)).Run()
```

This essentially (due to no fault of buildkite agent) means that the spawned process doesn't get any opportunity to perform any cleanup tasks. 

What do we do then if we want to run cleanup tasks when a job is cancelled?

**Allow custom cancel command**

Instead of using `signals` as the only supported mechanisms for indicating to the spawned process, we can allow a custom command to be executed. An example of a custom command may be invoking a HTTP API endpoint which will be exposed by the spawned process which will trigger the shutdown process as part of which it will perform cleanup tasks.

__How do we implement it?__

For example (to minimize changes across the board), the agent can look for a environment variable
`CANCEL_COMMAND`:

```
steps:

    - label:  ':hammer: Build and Run All Tests'
      command: 'powershell.exe -File .buildkite/scripts/RunTests.ps1'
      env:
          CANCEL_COMMAND: 'Invoke-WebRequest http://127.0.0.1:8000/cancel'
```

And then:

```
# process\process.go
func (p *Process) Kill() error {
       if os.Getenv("CANCEL_COMMAND") {
           # Execute the custom command
       } else {
          # Current signalling behavior
           # ..
	
    }
```

Of course what I propose here is going to be usable on non-Windows platforms as well. However, this would be useful most on Windows platforms. 

What do you folks think of this?

**Comment by lox** [comment]
Yeah, windows makes this really hard. If we had a `pre-cancel` hook, or had `pre-exit` reliably execute on windows, you could implement `CANCEL_COMMAND` yourself.

**Comment by amitsaha** [comment]
> pre-exit reliably execute on windows,

Could we make the `pre-exit` hook execute even when it's a cancellation?

**Comment by lox** [comment]
@amitsaha so currently we run `buildkite-agent bootstrap` as a sub-process to execute a job. It handles plugin setup, checkout and command execution, as well as all the hook.

We run `pre-exit` in that subprocess by catching the signal and running that hook before exiting, which is why it doesn't work in windows.

To make it work, we'd need to move `pre-exit` out of the bootstrap.

**Comment by amitsaha** [comment]
I can execute the pre-exit hook on Windows when the job exits on its own. I think i am a bit cofused by what you mean?

**Comment by lox** [comment]
My understanding is that there is no way in windows to catch a termination signal and execute some code before the process terminates. That is the mechanism with use in unix land to run the `pre-exit` hook before the bootstrap terminates when it cancels. Perhaps I'm misunderstanding what you are asking?

**Comment by amitsaha** [comment]
Ahh - no i misunderstood your previous comment about the non-reliability of pre-exit on Windows. I missed the part about it being the case you are referring to is the cancellation case and not the normal termination. 

Okay, yeah so any form of reliance on signal handling isn’t going to work on Windows.

**Comment by amitsaha** [comment]
https://github.com/buildkite/agent/pull/795 has definitely improved the behavior a lot.

**Comment by amitsaha** [comment]
I just wanted to write a bit more about my motivation in filing this issue and how I have worked around it.

Our CI setup is basically, we have a top-level powershell script that the pipeline runs. This powershell script then spawns other powershell scripts, which then spins up docker containers as part of build/testing. #795 helped here since now, we also have the entire process tree being terminated.

However, since on Windows, docker containers don't have their parents as the processes spawning them, they don't get killed. So, what do we do? Since we don't want running containers to hang around, we basically use the `pre-checkout` hook to always remove any running containers before starting a new build.

**Comment by lox** [comment]
Closed via #879 and #860.
